include/benchmark/benchmark.h

```diff
@@ -968,6 +969,25 @@ class Benchmark {
     return Ranges(ranges);
   }
 
+#if defined(BENCHMARK_HAS_CXX11)
+  // Have "setup" and/or "teardown" invoked once for every benchmark run.
+  // If the benchmark is multi-threaded (will run in k threads concurrently),
+  // the setup callback will be be invoked exactly once (not k times) before
+  // each run with k threads. Time allowing (e.g. for a short benchmark), there
+  // may be multiple such runs per benchmark, each run with its own
+  // "setup"/"teardown".
+  //
+  // If the benchmark uses different size groups of threads (e.g. via
+  // ThreadRange), the above will be true for each size group.
+  //
+  // The callback will be passed the number of threads for this benchmark run.
+  //
+  // The callback must not be self-deleting.  The Benchmark
+  // object takes ownership of the callback std::function object.
+  Benchmark* SetUp(std::function<void(benchmark::State&)>);
```

Comment:
Can you add motivational test why this can't just be a function pointer like see e.g. `BigOFunc`?